File: packages/shared/src/algorithms/geo.ts
/**
 * MODULE: @peapod/shared/algorithms/geo
 *
 * PURPOSE
 *   Spherical geometry and speed classification. The foundation every other
 *   location feature is built on: presence, geofencing, "together" detection,
 *   trip reconstruction, and the distance shown on a member's card.
 *
 * INPUTS  : coordinates in WGS84 decimal degrees; speeds in metres per second
 * OUTPUTS : distances in metres; labels as plain strings
 *
 * CONSUMED BY
 *   - services/api  : the TypeScript fallback for the Rust compute service
 *   - apps/mobile   : live distance readouts, which must update at frame rate
 *                     and so are computed locally rather than round-tripped
 *   (services/compute has the Rust twin of this module in src/geo.rs)
 *
 * ACCURACY NOTE
 *   Every function here models the Earth as a sphere of a single radius. That is
 *   off by up to about 0.5% versus a proper ellipsoidal calculation, because the
 *   Earth is flattened at the poles. For this product that error is irrelevant:
 *   the largest distance anything here decides on is a 220 m geofence, where
 *   0.5% is one metre -- far inside consumer GPS accuracy, which is 5-50 m on a
 *   good day. Using a Vincenty or Karney solution would add real complexity to
 *   buy precision that GPS noise would immediately swamp.
 */

import { GEO } from '../rules.js';
import type { ActivityClassification, ActivityLabel } from '../compute.js';

/** Degrees to radians. Hoisted so it is not recomputed inside the distance loops. */
const DEG_TO_RAD = Math.PI / 180;

/**
 * Great-circle distance between two coordinates.
 *
 * Uses the haversine formula, which stays numerically stable for the small
 * distances this app cares about. (The simpler spherical law of cosines loses
 * precision badly under about 1 km because it takes the arccosine of a value
 * extremely close to 1.)
 *
 * @param lat1 Latitude of the first point, decimal degrees.
 * @param lon1 Longitude of the first point, decimal degrees.
 * @param lat2 Latitude of the second point, decimal degrees.
 * @param lon2 Longitude of the second point, decimal degrees.
 * @returns Distance in METRES. Always non-negative.
 *
 * EDGE CASES
 *   - Identical points return exactly 0.
 *   - Handles the antimeridian correctly: the formula works on the difference of
 *     longitudes through a sine, so a jump from +179 to -179 degrees is treated
 *     as the 2-degree hop it really is, not a 358-degree one.
 */
export function haversine(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const dLat = (lat2 - lat1) * DEG_TO_RAD;
  const dLon = (lon2 - lon1) * DEG_TO_RAD;

  // `a` is the square of half the chord length between the points.
  const sinHalfLat = Math.sin(dLat / 2);
  const sinHalfLon = Math.sin(dLon / 2);
  const a =
    sinHalfLat * sinHalfLat +
    Math.cos(lat1 * DEG_TO_RAD) * Math.cos(lat2 * DEG_TO_RAD) * sinHalfLon * sinHalfLon;

  // atan2 rather than asin: atan2 stays accurate as `a` approaches 1
  // (antipodal points), where asin would lose all precision.
  return GEO.earthRadius_m * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

/**
 * Finds the first saved place a location falls inside.
 *
 * Returns the FIRST match rather than the closest, matching the original
 * behaviour. That is intentional and worth keeping: saved places are ordered
 * with the most significant first (home, work), and when two overlap -- a cafe
 * inside a mall, say -- the earlier one is the more meaningful label. Picking
 * the geometrically nearest would flip the label back and forth as GPS drifted
 * between two nearby places.
 *
 * @param location The member's current position.
 * @param places   Candidate saved places, in significance order.
 * @param radius_m How close counts as "at" the place. Defaults to the shared
 *                 place radius, which equals the geofence entry radius so the
 *                 two can never disagree.
 * @returns The matching place and its distance in metres, or null when the
 *          member is not at any saved place.
 */
export function nearestPlace<T extends { latitude: number; longitude: number }>(
  location: { latitude: number; longitude: number } | null | undefined,
  places: readonly T[] | null | undefined,
  radius_m: number = GEO.placeRadius_m,
): { place: T; distance_m: number } | null {
  if (!location || !places || places.length === 0) return null;

  for (const place of places) {
    const distance_m = haversine(location.latitude, location.longitude, place.latitude, place.longitude);
    if (distance_m <= radius_m) return { place, distance_m };
  }

  return null;
}

/**
 * Renders a distance for display.
 *
 * The precision deliberately coarsens with distance, because precision the GPS
 * cannot support reads as false confidence. "1.2 km" is honest; "1,237 m" is
 * not, when the fix itself is +/- 20 m.
 *
 * @param metres Distance in metres, or null/undefined when unknown.
 * @returns A short label: `"Together"`, `"420 m"`, `"1.2 km"`, `"38 km"`, or an
 *          em dash when the distance is unknown.
 */
export function formatDistance(metres: number | null | undefined): string {
  if (metres == null) return '—';
  // Below the together threshold the exact number is noise, and the social fact
  // ("you are in the same place") is what the member wants to read.
  if (metres <= GEO.togetherThreshold_m) return 'Together';
  if (metres < 1000) return `${Math.round(metres)} m`;
  if (metres < 10000) return `${(metres / 1000).toFixed(1)} km`;
  return `${Math.round(metres / 1000)} km`;
}

/**
 * Converts metres per second to whole kilometres per hour.
 *
 * @param mps Speed in metres per second. Null and undefined are treated as 0.
 * @returns Speed in km/h, rounded to an integer.
 */
export function toKmh(mps: number | null | undefined): number {
  return Math.round((mps ?? 0) * 3.6);
}

/**
 * Whether a speed reading looks like vehicle travel.
 *
 * @param mps Speed in metres per second.
 * @returns True above roughly 11 km/h.
 */
export function isDriving(mps: number | null | undefined): boolean {
  return (mps ?? 0) > GEO.drivingThreshold_mps;
}

/**
 * Classifies a speed reading into a human activity label.
 *
 * Band boundaries, from `data/rules/geo.json`:
 *   < 0.6 m/s  (~2 km/h)   Stationary  -- standing still, or pure GPS noise
 *   < 1.6 m/s  (~6 km/h)   Walking
 *   < 3.0 m/s  (~11 km/h)  Cycling     -- or a brisk jog
 *   >= 3.0 m/s             Driving
 *
 * @param mps Speed in metres per second.
 * @returns The label plus a driving flag, so callers do not re-derive it.
 *
 * KNOWN LIMITATION
 *   This is best-effort. Neither mobile platform gives this layer the
 *   accelerometer-based activity recognition that would distinguish a passenger
 *   on a bus from a cyclist, so a car stopped at a red light reads as
 *   "Stationary" and a fast downhill cyclist reads as "Driving".
 */
export function classifyActivity(mps: number | null | undefined): ActivityClassification {
  const speed = mps ?? 0;

  let label: ActivityLabel;
  if (speed < GEO.stationaryBelow_mps) label = 'Stationary';
  else if (speed < GEO.walkingBelow_mps) label = 'Walking';
  else if (speed < GEO.cyclingBelow_mps) label = 'Cycling';
  else label = 'Driving';

  return { label, driving: label === 'Driving' };
}

/**
 * Reduces a dense polyline to distance-spaced waypoints.
 *
 * WHY THIS EXISTS
 *   Road-snapping routers behave badly at both extremes. Feed one every raw GPS
 *   fix and it U-turns onto noisy mid-block points, drawing the same street
 *   twice. Feed it only origin and destination and it collapses a real revisit
 *   into a single pass, so a street the member genuinely drove down twice is
 *   drawn once. Spacing waypoints a few hundred metres apart follows the actual
 *   path while staying sparse enough to avoid the phantom U-turns.
 *
 * @param points   Ordered `[latitude, longitude]` pairs.
 * @param minGap_m Minimum spacing between kept waypoints, in metres.
 * @returns A subset of the input, always including the first and last points so
 *          the origin pin and destination stay exactly where the member was.
 */
export function downsampleByDistance(
  points: readonly [number, number][],
  minGap_m: number = GEO.routerWaypointGap_m,
): [number, number][] {
  if (points.length <= 2) return [...points];

  const first = points[0]!;
  const out: [number, number][] = [first];
  let last = first;

  // Interior points only -- the final point is appended unconditionally below.
  for (let i = 1; i < points.length - 1; i++) {
    const candidate = points[i]!;
    if (haversine(last[0], last[1], candidate[0], candidate[1]) >= minGap_m) {
      out.push(candidate);
      last = candidate;
    }
  }

  out.push(points[points.length - 1]!);
  return out;
}

/**
 * Whether two members are close enough to render as "Together".
 *
 * @param a First member's position.
 * @param b Second member's position.
 * @returns True within the together threshold (100 m). False when either
 *          position is unknown -- absence of data is not evidence of proximity.
 */
export function areTogether(
  a: { latitude: number; longitude: number } | null | undefined,
  b: { latitude: number; longitude: number } | null | undefined,
): boolean {
  if (!a || !b) return false;
  return haversine(a.latitude, a.longitude, b.latitude, b.longitude) <= GEO.togetherThreshold_m;
}

/**
 * Decides whether a geofence state should flip, using hysteresis.
 *
 * WHY TWO RADII
 *   With a single boundary, a member sitting near the edge of their geofence
 *   fires arrived/left/arrived/left forever as GPS wanders a few metres, and
 *   every flip notifies the whole pod. Entry uses the tighter radius (150 m) and
 *   exit the looser one (220 m), so leaving requires travelling meaningfully
 *   further than arriving did. The 70 m gap has to be crossed before the state
 *   can change back.
 *
 * @param distance_m Current distance from the place's centre, in metres.
 * @param wasInside  Whether the member was considered inside on the previous fix.
 * @returns True when the member should now be considered inside the fence.
 */
export function isInsideGeofence(distance_m: number, wasInside: boolean): boolean {
  return wasInside ? distance_m <= GEO.geofenceExit_m : distance_m <= GEO.geofenceEnter_m;
}

File: services/compute/src/geo.rs
//! # Geo primitives
//!
//! ## Purpose
//! The distance and speed maths every location feature is built on. This is a
//! faithful port of the original `src/lib/geo.js`; the thresholds it compares
//! against are read from `packages/shared/data/rules/geo.json` via
//! `crate::rules::GEO` rather than re-typed as literals.
//!
//! ## Inputs
//! Bare coordinates in WGS84 degrees, speeds in metres per second, and lists of
//! saved places. Nothing here is stateful and nothing reads a clock.
//!
//! ## Outputs
//! Distances in **metres** (always metres inside this module -- only
//! `format_distance` ever produces kilometres, and only as display text),
//! human-readable labels, and place matches.
//!
//! ## Who calls this
//! - `crate::trips`, on every ping pair of the reconstruction pipeline. This is
//!   the hottest code in the service, which is why `haversine` takes plain
//!   `f64`s rather than a struct.
//! - `POST /geo/nearest-place` and `POST /geo/activity`, called by
//!   `services/api` when rendering a member's presence card.

use crate::models::{ActivityClassification, GeoPoint, PlaceInput};
use crate::rules::GEO;

/// Great-circle distance between two points, in **metres**.
///
/// # Parameters
/// - `lat1`, `lon1`: first point, WGS84 degrees.
/// - `lat2`, `lon2`: second point, WGS84 degrees.
///
/// # Returns
/// Metres along the surface of a sphere of radius
/// `geo.json -> earthRadiusM` (6 371 000 m, the mean Earth radius).
///
/// # Algorithm (haversine), step by step
/// 1. Convert the latitude and longitude *differences* to radians.
/// 2. `a` = the square of half the chord length between the points:
///    `sin²(Δlat/2) + cos(lat1)·cos(lat2)·sin²(Δlon/2)`.
/// 3. The central angle is `2·atan2(√a, √(1-a))`. Using `atan2` rather than
///    `asin(√a)` keeps the result numerically stable for antipodal points.
/// 4. Multiply the angle (radians) by the radius to get metres.
///
/// # Edge cases
/// - Identical points return exactly `0.0`.
/// - A spherical Earth is an approximation; error against WGS84 ellipsoidal
///   distance is well under 0.5%, which is far smaller than the GPS accuracy
///   this service filters on, so it is irrelevant here.
/// - `NaN` inputs propagate as `NaN`. The caller (`services/api`) validates
///   coordinates before storing pings, so this is not defended against.
pub fn haversine(lat1: f64, lon1: f64, lat2: f64, lon2: f64) -> f64 {
    let radius_m = GEO.earth_radius_m;
    let d_lat = (lat2 - lat1).to_radians();
    let d_lon = (lon2 - lon1).to_radians();
    let a = (d_lat / 2.0).sin().powi(2)
        + lat1.to_radians().cos() * lat2.to_radians().cos() * (d_lon / 2.0).sin().powi(2);
    radius_m * 2.0 * a.sqrt().atan2((1.0 - a).sqrt())
}

/// Human-readable distance text.
///
/// # Parameters
/// - `metres`: `None` when the distance is unknown (e.g. the other member has
///   never pinged).
///
/// # Returns
/// - `"—"` (an em dash) when unknown.
/// - `"Together"` at or below `geo.json -> proximity.togetherThreshold_m`
///   (100 m). The `why`: "Roughly the width of a building, so being in the same
///   cafe reads as together despite GPS scatter."
/// - `"NNN m"` below 1 km, rounded to the metre.
/// - `"N.N km"` below 10 km, one decimal.
/// - `"NN km"` at 10 km and above, rounded to the kilometre -- one decimal of a
///   double-digit kilometre figure is noise, not information.
///
/// # Edge cases
/// Rounding uses Rust's round-half-away-from-zero, whereas JavaScript's
/// `Math.round` is round-half-up. They differ only for negative halves, and a
/// distance is never negative, so the two implementations agree.
pub fn format_distance(metres: Option<f64>) -> String {
    let m = match metres {
        Some(m) => m,
        None => return "—".to_string(),
    };
    if m <= GEO.proximity.together_threshold_m.value {
        return "Together".to_string();
    }
    if m < 1000.0 {
        return format!("{} m", m.round() as i64);
    }
    if m < 10_000.0 {
        return format!("{:.1} km", m / 1000.0);
    }
    format!("{} km", (m / 1000.0).round() as i64)
}

/// Metres per second to whole kilometres per hour.
///
/// # Parameters
/// - `metres_per_second`: as reported by the device.
///
/// # Returns
/// `round(m/s × 3.6)` as an integer. Whole km/h because the UI shows a badge,
/// not an instrument reading.
pub fn to_kmh(metres_per_second: f64) -> i64 {
    (metres_per_second * 3.6).round() as i64
}

/// Turns a speed reading into a human activity label.
///
/// # Parameters
/// - `speed_ms`: metres per second. A missing reading arrives as `0.0` (see
///   `PingInput`'s `#[serde(default)]`), which classifies as `Stationary`,
///   exactly like the original JavaScript's `speedMs || 0`.
///
/// # Returns
/// One of `Stationary`, `Walking`, `Cycling`, `Driving`, plus a `driving` flag
/// the map uses to decide whether to road-snap a line.
///
/// # Bands (all from `geo.json -> activityBands`)
/// - `< 0.6 m/s` -> Stationary. "Under about 2 km/h -- standing still, or GPS
///   noise."
/// - `< 1.6 m/s` -> Walking. "About 2 to 6 km/h -- walking pace."
/// - `< 3.0 m/s` -> Cycling. "About 6 to 11 km/h -- cycling or a brisk jog."
/// - `>= 3.0 m/s` -> Driving. "Above about 11 km/h is treated as driving."
///
/// # Edge cases
/// The bands are strictly increasing and checked in order, so a value exactly
/// on a boundary falls into the *higher* band (0.6 m/s is Walking, not
/// Stationary) -- identical to the original's `<` comparisons.
pub fn classify_activity(speed_ms: f64) -> ActivityClassification {
    let bands = &GEO.activity_bands;
    if speed_ms < bands.stationary_below_mps.value {
        return ActivityClassification { label: "Stationary", driving: false };
    }
    if speed_ms < bands.walking_below_mps.value {
        return ActivityClassification { label: "Walking", driving: false };
    }
    if speed_ms < bands.cycling_below_mps.value {
        return ActivityClassification { label: "Cycling", driving: false };
    }
    ActivityClassification { label: "Driving", driving: true }
}

/// Whether a speed reading counts as driving on its own.
///
/// # Parameters
/// - `speed_ms`: metres per second.
///
/// # Returns
/// `true` above `geo.json -> activityBands.drivingThreshold_mps` (3.0 m/s,
/// about 11 km/h). Strictly greater-than, matching the original `> 3`.
pub fn is_driving(speed_ms: f64) -> bool {
    speed_ms > GEO.activity_bands.driving_threshold_mps.value
}

/// Finds the saved place a member counts as being "at".
///
/// # Parameters
/// - `location`: where the member is now.
/// - `places`: candidate places **in the caller's order** -- see the note below.
/// - `radius_m`: metres. `POST /geo/nearest-place` defaults this to
///   `geo.json -> proximity.placeRadius_m` (150 m), which is also the geofence
///   entry radius, "so 'at a place' and 'inside the fence' can never disagree".
///
/// # Returns
/// `Some((place_id, distance_m))` for the match, or `None` when the member is
/// outside every place's radius.
///
/// # Important: this returns the FIRST match, not the closest
/// The original `nearestPlace()` returned the first place within the radius and
/// this port keeps that behaviour deliberately, because the caller relies on it:
/// places are passed in the app's own priority order (a timed/temporary place
/// before a permanent one), so "first within the radius" is a *preference*
/// order, not an accident. Picking the geometrically nearest place instead would
/// silently change which place a member appears to be at whenever two overlap.
/// The distance returned is the distance to that first match.
///
/// # Edge cases
/// - An empty `places` list returns `None`.
/// - A non-positive `radius_m` returns `None` for every place, since the
///   comparison is `<=` against a distance that is `>= 0`; only a member sitting
///   exactly on a place would match at radius 0. The handler rejects negative
///   radii outright.
pub fn nearest_place(
    location: &GeoPoint,
    places: &[PlaceInput],
    radius_m: f64,
) -> Option<(String, f64)> {
    for place in places {
        let distance_m = haversine(
            location.latitude,
            location.longitude,
            place.latitude,
            place.longitude,
        );
        if distance_m <= radius_m {
            return Some((place.id.clone(), distance_m));
        }
    }
    None
}

/// Reduces a polyline to distance-spaced waypoints.
///
/// # Parameters
/// - `points`: ordered `[latitude, longitude]` pairs.
/// - `min_gap_m`: minimum spacing in metres. Callers pass
///   `geo.json -> tripReconstruction.routerWaypointGap_m` (250 m).
///
/// # Returns
/// A new vector that always begins with the first point and ends with the last,
/// with interior points thinned so consecutive kept points are at least
/// `min_gap_m` apart along the path.
///
/// # Why 250 m
/// From the rule file: "Feeding every raw fix makes the router U-turn onto noisy
/// mid-block points and draw the same road twice; routing only origin to
/// destination collapses real revisits into a single pass. This spacing follows
/// the actual path while staying sparse enough to avoid both failures."
///
/// # Algorithm
/// Walk the interior points, keeping a running `last_kept`. A point is kept when
/// it is at least `min_gap_m` from `last_kept` (not from its predecessor), so a
/// dense cluster collapses to one waypoint instead of ratcheting forward.
///
/// # Edge cases
/// Inputs of 2 points or fewer are returned unchanged -- there is no interior to
/// thin, and both endpoints must survive.
pub fn downsample_by_distance(points: &[[f64; 2]], min_gap_m: f64) -> Vec<[f64; 2]> {
    if points.len() <= 2 {
        return points.to_vec();
    }
    let mut out: Vec<[f64; 2]> = Vec::with_capacity(points.len());
    out.push(points[0]);
    let mut last_kept = points[0];
    // `points.len() - 1` is safe: len > 2 here. The last point is appended
    // unconditionally after the loop, so the loop stops one short of it.
    for point in &points[1..points.len() - 1] {
        if haversine(last_kept[0], last_kept[1], point[0], point[1]) >= min_gap_m {
            out.push(*point);
            last_kept = *point;
        }
    }
    out.push(points[points.len() - 1]);
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    /// One degree of longitude at the equator on a sphere of radius R is
    /// `R · π/180`. With R = 6 371 000 m that is 111 194.93 m, which is a
    /// closed-form check on the whole formula rather than on a memorised number.
    #[test]
    fn haversine_matches_the_closed_form_at_the_equator() {
        let expected = 6_371_000.0 * std::f64::consts::PI / 180.0;
        let d = haversine(0.0, 0.0, 0.0, 1.0);
        assert!(
            (d - expected).abs() < 0.001,
            "one degree of longitude at the equator: got {d}, expected {expected}"
        );
        // A degree of latitude is the same arc on a sphere.
        let d_lat = haversine(0.0, 0.0, 1.0, 0.0);
        assert!((d_lat - expected).abs() < 0.001, "one degree of latitude: got {d_lat}");
    }

    /// A known real-world pair: London (51.5007, -0.1246) to Paris
    /// (48.8567, 2.3508) is roughly 343 km great-circle.
    #[test]
    fn haversine_matches_a_known_city_pair() {
        let d = haversine(51.5007, -0.1246, 48.8567, 2.3508);
        assert!(
            d > 340_000.0 && d < 345_000.0,
            "London to Paris should be ~343 km, got {d} m"
        );
    }

    #[test]
    fn haversine_is_zero_for_identical_points_and_symmetric() {
        assert_eq!(haversine(1.3521, 103.8198, 1.3521, 103.8198), 0.0);
        let forward = haversine(1.3521, 103.8198, 35.6762, 139.6503);
        let backward = haversine(35.6762, 139.6503, 1.3521, 103.8198);
        assert!((forward - backward).abs() < 1e-6);
    }

    #[test]
    fn format_distance_covers_every_band() {
        assert_eq!(format_distance(None), "—");
        // At or below the 100 m together threshold.
        assert_eq!(format_distance(Some(0.0)), "Together");
        assert_eq!(format_distance(Some(100.0)), "Together");
        // Metres.
        assert_eq!(format_distance(Some(100.4)), "100 m");
        assert_eq!(format_distance(Some(999.6)), "1000 m");
        // Sub-10 km, one decimal.
        assert_eq!(format_distance(Some(1000.0)), "1.0 km");
        assert_eq!(format_distance(Some(5432.0)), "5.4 km");
        // Double-digit kilometres, whole numbers.
        assert_eq!(format_distance(Some(10_000.0)), "10 km");
        assert_eq!(format_distance(Some(15_400.0)), "15 km");
    }

    #[test]
    fn to_kmh_rounds_to_whole_units() {
        assert_eq!(to_kmh(0.0), 0);
        assert_eq!(to_kmh(1.0), 4); // 3.6 -> 4
        assert_eq!(to_kmh(10.0), 36);
        assert_eq!(to_kmh(27.777), 100);
    }

    #[test]
    fn activity_bands_match_the_rule_file() {
        assert_eq!(classify_activity(0.0).label, "Stationary");
        assert_eq!(classify_activity(0.59).label, "Stationary");
        // Boundaries fall into the higher band, as in the original.
        assert_eq!(classify_activity(0.6).label, "Walking");
        assert_eq!(classify_activity(1.59).label, "Walking");
        assert_eq!(classify_activity(1.6).label, "Cycling");
        assert_eq!(classify_activity(2.99).label, "Cycling");
        assert_eq!(classify_activity(3.0).label, "Driving");
        assert!(classify_activity(3.0).driving);
        assert!(!classify_activity(2.9).driving);
    }

    #[test]
    fn is_driving_is_strictly_above_the_threshold() {
        assert!(!is_driving(3.0));
        assert!(is_driving(3.01));
    }

    #[test]
    fn nearest_place_returns_the_first_match_inside_the_radius() {
        let here = GeoPoint { latitude: 1.3000, longitude: 103.8000 };
        // ~11 m away: comfortably inside 150 m.
        let close = PlaceInput { id: "home".into(), latitude: 1.3001, longitude: 103.8000 };
        // ~11 km away: outside.
        let far = PlaceInput { id: "office".into(), latitude: 1.4000, longitude: 103.8000 };

        let hit = nearest_place(&here, &[close.clone(), far.clone()], 150.0);
        let (id, distance) = hit.expect("the close place should match");
        assert_eq!(id, "home");
        assert!(distance < 150.0);

        assert!(nearest_place(&here, &[far.clone()], 150.0).is_none());
        assert!(nearest_place(&here, &[], 150.0).is_none());

        // Order is a preference order: when two places both contain the member,
        // the earlier one wins even if the later one is nearer.
        let nearer = PlaceInput { id: "nearer".into(), latitude: 1.30000, longitude: 103.80000 };
        let hit = nearest_place(&here, &[close, nearer], 150.0);
        assert_eq!(hit.expect("first match wins").0, "home");
    }

    #[test]
    fn downsample_keeps_endpoints_and_thins_the_middle() {
        // Five points about 11 m apart. With a 250 m gap every interior point is
        // dropped and only the endpoints survive.
        let dense: Vec<[f64; 2]> = (0..5).map(|i| [1.3 + (i as f64) * 0.0001, 103.8]).collect();
        let out = downsample_by_distance(&dense, 250.0);
        assert_eq!(out.len(), 2);
        assert_eq!(out[0], dense[0]);
        assert_eq!(out[1], dense[4]);

        // With a tiny gap, every point survives.
        let out = downsample_by_distance(&dense, 1.0);
        assert_eq!(out.len(), 5);

        // Two points or fewer pass straight through.
        assert_eq!(downsample_by_distance(&dense[..2], 250.0).len(), 2);
        assert_eq!(downsample_by_distance(&[], 250.0).len(), 0);
    }
}
